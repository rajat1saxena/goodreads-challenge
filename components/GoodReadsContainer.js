/**
 * This is the master component which houses all GoodReads related
 * functionalities.
 */
import React from 'react'
import SearchArea from './SearchArea'
import HeroArea from './HeroArea'

const GoodReadsContainer = (props) => (
  <div className="GoodReadsContainer">
    <SearchArea />
    <HeroArea />
    <style jsx>{`
      .GoodReadsContainer {
        display: flex;
        flex-direction: row;
        border: 1px solid #eee;
        padding: 5px;
        border-radius: 5px;
      }
    `}</style>
  </div>
)

export default GoodReadsContainer
