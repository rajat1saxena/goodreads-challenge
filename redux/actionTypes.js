/**
 * This file defines all the expected action types.
 */

export const QUERY_CHANGED = "QUERY_CHANGED"
export const INITIATE_SEARCH = "INITIATE_SEARCH"
export const SEARCH_SUCCESS = "SEARCH_SUCCESS"
export const SEARCH_FAILED = "SEARCH_FAILED"
export const LOAD_BOOK_DETAILS = "LOAD_BOOK_DETAILS"
export const LOAD_BOOK_DETAILS_SUCCESS = "LOAD_BOOK_DETAILS_SUCCESS"
export const LOAD_BOOK_FAILED = "LOAD_BOOK_FAILED"