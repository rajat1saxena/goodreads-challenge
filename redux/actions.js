/**
 * This file contains action creators for all state related functionality.
 */

import {
    QUERY_CHANGED,
    SEARCHING_QUERY,
    SEARCHING_QUERY_SUCCESS,
    LOADING_BOOK_DETAILS,
    LOADING_BOOK_DETAILS_SUCCESS
} from './actionTypes.js'

export function queryChanged(newQuery) {
    return { type: QUERY_CHANGED, newQuery: newQuery }
}

export function searchBooks() {
    
}

export function fetchBookDetailsById(bookId) {
    return { type: FETCH_DETAILS, id: bookId }
}