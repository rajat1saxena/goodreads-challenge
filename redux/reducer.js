/**
 * The reducer for changing the app state
 */

import {
    QUERY_CHANGED,
    INITIATE_SEARCH,
    SEARCH_SUCCESS,
    SEARCH_FAILED,
    LOAD_BOOK_FAILED,
    LOAD_BOOK_DETAILS,
    LOAD_BOOK_DETAILS_SUCCESS
} from './actionTypes.js'
import BookItem from '../lib/BookItem.js';

// This represents the initial state our app will be in 
// when it is loaded for the very first time.
const initialState = {
    query: '',
    loadingSearch: false,
    loadingBookDetails: false,
    lastSearchResults: [],
    bookDetails: null
}

// The reducer function
export default function (state = initialState, action) {
    // console.log(action)
    switch(action.type) {
        case QUERY_CHANGED:
            return Object.assign({}, state, {query: action.newQuery})
            break
        case INITIATE_SEARCH:
            return Object.assign({}, state, {loadingSearch: true})
            break
        case SEARCH_SUCCESS:
            const resultItems = []
            const result = JSON.parse(action.data)
            for (let i of result.GoodreadsResponse.search.results.work) {
              resultItems.push(
                new BookItem(
                  i.best_book.id._text,
                  i.best_book.title._text,
                  i.best_book.author.name._text,
                  i.ratings_count._text,
                  i.average_rating._text,
                  i.best_book.small_image_url._text
                )
              )
            }

            return Object.assign({}, state, {
              loadingSearch: false,
              lastSearchResults: resultItems
            })
            break
        case SEARCH_FAILED:
            return Object.assign({}, state, {loadingSearch: false})
            break
        case LOAD_BOOK_DETAILS:
            return Object.assign({}, state, {loadingBookDetails: true})
            break
        case LOAD_BOOK_FAILED:
            return Object.assign({}, state, {loadingBookDetails: false})
            break
        case LOAD_BOOK_DETAILS_SUCCESS:
            const jsonData = JSON.parse(action.data)
            const book = jsonData.GoodreadsResponse.book
            const id = book.id._text || ''
            const title = book.title._text || book.title._cdata || ''
            const author = (book.authors.author.length > 1 ? 
              book.authors.author[0].name._text 
              : 
              book.authors.author.name._text) || ''
            const ratings_count = book.ratings_count._text || book.ratings_count._cdata || ''
            const average_rating = book.average_rating._text || book.average_rating._cdata || ''
            const image = book.image_url._text || book.image_url._cdata || ''
            const description = book.description._text || book.description._cdata || ''
            const bookItem = new BookItem(
              id,
              title,
              author,
              ratings_count,
              average_rating,
              image,
              description
            )
            
            return Object.assign({}, state, {
              loadingBookDetails: false,
              bookDetails: bookItem
            })
            break
        default:
            return state
    }
}

