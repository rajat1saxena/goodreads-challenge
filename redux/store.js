/**
 * A redux store for the app
 */
import { createStore } from 'redux'
import reducer from './reducer.js'

const store = createStore(reducer)
store.subscribe(() => {
    console.log(store.getState())
})
export default store