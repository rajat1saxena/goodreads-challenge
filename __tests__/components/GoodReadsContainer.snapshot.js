import React from 'react'
import renderer from 'react-test-renderer'
import GoodReadsContainer from '../../components/GoodReadsContainer.js'
import { createStore } from 'redux'
import { Provider } from 'react-redux'
import { 
  initialState 
} from '../../redux/reducer.js'

it('Should render correctly', () => {
  const store = createStore((state = initialState) => state)
  const tree = renderer.create(
      <Provider store={store}>
        <GoodReadsContainer />
      </Provider>
  ).toJSON()
  expect(tree).toMatchSnapshot()
})