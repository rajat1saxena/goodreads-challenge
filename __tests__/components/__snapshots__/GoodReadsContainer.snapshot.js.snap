// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render correctly 1`] = `
<div
  className="jsx-3615494824 GoodReadsContainer"
>
  <div
    className="jsx-2760136266 SearchArea"
  >
    <div
      className="jsx-4012630625 container"
    >
      <input
        className="jsx-4012630625"
        onChange={[Function]}
        placeholder="Search GoodReads.com books"
        type="text"
        value=""
      />
      <button
        className="jsx-4012630625"
        data-testid="submit"
        onClick={[Function]}
        type="submit"
      >
        Search
      </button>
    </div>
    <div
      className="jsx-1729626677"
    >
      <div
        className="jsx-1729626677 Message"
      >
        Nothing to show here.
      </div>
    </div>
  </div>
  <div
    className="jsx-4043294663 HeroArea"
  >
    <div
      className="jsx-4043294663 Message"
    >
      Click on a book to see details.
    </div>
  </div>
</div>
`;
